### PR TITLE
Json SBOM improvements

### PR DESCRIPTION
**Pull request by m-1-k-3**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Feature

* **What is the current behavior?** (You can also link to an open issue here)

duplicates
long time running

* **What is the new behavior (if this is a feature change)? If possible add a screenshot.**

better duplicate handling
s09 also uses package-manager in sbom-mode and drops tests on bins handled by package manager. This results in a massive performance boost. On some firmwares we decreased our testing time from more than 4h to 1h without loosing results.
split s08 in sub-modules
deep extractor can be limited by scanning profile (see sbom-default as example)
s08 sub-modules can be configured via scanning profiles (see sbom-default as example)

* **Other information**:

Most of the time the sbom-default is quite fine to use. If you do not need binary processing at all you can switch to sbom-minimal

### CODE REVIEW

File: modules/S08_main_package_sbom_modules/S08_submodule_deb_package_parser.sh [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+      lAPP_VENDOR="${lAPP_NAME}"
+      lCPE_IDENTIFIER="cpe:${CPE_VERSION}:a:${lAPP_VENDOR}:${lAPP_NAME}:${lAPP_VERS}:*:*:*:*:*:*"
+
+      if [[ -z "${lOS_IDENTIFIED}" ]]; then
+        lOS_IDENTIFIED="debian-based"
+      fi
+      lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "deb" "${lAPP_NAME:-NA}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+      local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS:-NA}"
+
+      if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_deb_package_parser.sh:138]: That not a dependency?

File: modules/S08_main_package_sbom_modules/S08_submodule_debian_pkg_mgmt_parser.sh [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+  local lAPP_VENDOR="NA"
+  local lCPE_IDENTIFIER="NA"
+  local lMD5_CHECKSUM="NA"
+  local lSHA256_CHECKSUM="NA"
+  local lSHA512_CHECKSUM="NA"
+  local lPURL_IDENTIFIER="NA"
+  local lAPP_DEPS=""
+  local lAPP_DEPS_ARR=()
+
+  # Package: dbus - Status: install ok installed - Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com> - Version: 1.12.16-2ubuntu2.1 - Description: simple interprocess messaging system (daemon and utilities)
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_debian_pkg_mgmt_parser.sh:115]: Guessing that's an example input

File: modules/S08_main_package_sbom_modules/S08_submodule_java_archives_parser.sh [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+      lAPP_VENDOR="${lAPP_NAME}"
+      lCPE_IDENTIFIER="cpe:${CPE_VERSION}:a:${lAPP_VENDOR}:${lAPP_NAME}:${lAPP_VERS}:*:*:*:*:*:*"
+
+      if [[ -z "${lOS_IDENTIFIED}" ]]; then
+        lOS_IDENTIFIED="generic"
+      fi
+      lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "java" "${lAPP_NAME:-NA}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+      local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS:-NA}"
+
+      if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_java_archives_parser.sh:116]: There a dep checker for jo?

File: modules/S08_main_package_sbom_modules/S08_submodule_nodejs_pcklockparser.sh [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+  lAPP_VENDOR="${lAPP_NAME}"
+  lCPE_IDENTIFIER="cpe:${CPE_VERSION}:a:${lAPP_VENDOR}:${lAPP_NAME}:${lAPP_VERS}:*:*:*:*:*:*"
+
+  if [[ -z "${lOS_IDENTIFIED}" ]]; then
+    lOS_IDENTIFIED="generic"
+  fi
+  lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "npm" "${lAPP_NAME:-NA}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+  local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS:-NA}"
+
+  if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_nodejs_pcklockparser.sh:130]: Same here

File: modules/S08_main_package_sbom_modules/S08_submodule_openwrt_pkg_mgmt_parser.sh [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        lAPP_VENDOR="${lAPP_NAME}"
+        lCPE_IDENTIFIER="cpe:${CPE_VERSION}:a:${lAPP_VENDOR}:${lAPP_NAME}:${lAPP_VERS}:*:*:*:*:*:*"
+
+        if [[ -z "${lOS_IDENTIFIED}" ]]; then
+          lOS_IDENTIFIED="openwrt"
+        fi
+        lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "opkg" "${lAPP_NAME:-NA}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+        local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS}"
+
+        if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_openwrt_pkg_mgmt_parser.sh:102]: Same question

File: modules/S08_main_package_sbom_modules/S08_submodule_python_pip_package_mgmt_parser.sh [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+        lAPP_VENDOR="${lAPP_NAME}"
+        lCPE_IDENTIFIER="cpe:${CPE_VERSION}:a:${lAPP_VENDOR}:${lAPP_NAME}:${lAPP_VERS}:*:*:*:*:*:*"
+
+        if [[ -z "${lOS_IDENTIFIED}" ]]; then
+          lOS_IDENTIFIED="generic"
+        fi
+        lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "pypi" "${lAPP_NAME:-NA}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+        local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS:-NA}"
+
+        if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_python_pip_package_mgmt_parser.sh:101]: .

File: modules/S08_main_package_sbom_modules/S08_submodule_python_poetry_lock_parser.sh [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        if [[ -z "${lOS_IDENTIFIED}" ]]; then
+          lOS_IDENTIFIED="generic"
+        fi
+        lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "pypi" "${lAPP_NAME:-NA}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+
+        local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS:-NA}"
+
+        # Todo: checksum
+
+        if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_python_poetry_lock_parser.sh:125]: .

File: modules/S08_main_package_sbom_modules/S08_submodule_python_requirements_parser.sh [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+  local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS:-NA}"
+
+        # with internet we can query further details
+        # └─$ curl -sH "accept: application/json" https://pypi.org/pypi/"${lAPP_NAME}"/json | jq '.info.author, .info.classifiers'
+        # "Chris P"
+        # [
+        #   "Development Status :: 5 - Production/Stable",
+        #   "License :: OSI Approved :: MIT License",
+
+  if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_python_requirements_parser.sh:164]: .

File: modules/S08_main_package_sbom_modules/S08_submodule_rpm_package_parser.sh [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        lOS_IDENTIFIED="rpm-based"
+      fi
+      lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "rpm" "${lAPP_NAME:-NA}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+
+      local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS:-NA}"
+
+      mapfile -t lRPM_FILES_ARR < <(rpm -qlp "${lRPM_ARCHIVE}" 2>/dev/null || true)
+      mapfile -t lRPM_DEPS_ARR < <(rpm -qR "${lRPM_ARCHIVE}" 2>/dev/null || true)
+
+      if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_rpm_package_parser.sh:119]: .

File: modules/S08_main_package_sbom_modules/S08_submodule_rpm_pkg_mgmt_parser.sh [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        lOS_IDENTIFIED="rpm-based"
+      fi
+      lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "rpm" "${lAPP_NAME:-NA}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+
+      local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS:-NA}"
+
+      mapfile -t lRPM_FILES_ARR < <(rpm -qlp "${lRPM_ARCHIVE}" 2>/dev/null || true)
+      mapfile -t lRPM_DEPS_ARR < <(rpm -qR "${lRPM_ARCHIVE}" 2>/dev/null || true)
+
+      if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_rpm_pkg_mgmt_parser.sh:119]: .

File: modules/S08_main_package_sbom_modules/S08_submodule_rust_cargo_lock_parser.sh [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        lCPE_IDENTIFIER="cpe:${CPE_VERSION}:a:${lAPP_VENDOR}:${lAPP_NAME}:${lAPP_VERS}:*:*:*:*:*:*"
+
+        if [[ -z "${lOS_IDENTIFIED}" ]]; then
+          lOS_IDENTIFIED="generic"
+        fi
+        lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "cargo" "${lAPP_NAME:-NA}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+
+        local lSTRIPPED_VERSION="::${lAPP_NAME}:${lAPP_VERS:-NA}"
+
+        if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_rust_cargo_lock_parser.sh:119]: .

File: modules/S08_main_package_sbom_modules/S08_submodule_windows_exifparser.sh [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+  if [[ -z "${lOS_IDENTIFIED}" ]]; then
+    lOS_IDENTIFIED="windows-based"
+  fi
+
+  lPURL_IDENTIFIER=$(build_purl_identifier "${lOS_IDENTIFIED:-NA}" "exe" "${lAPP_NAME//\.exe}" "${lAPP_VERS:-NA}" "${lAPP_ARCH:-NA}")
+
+  local lSTRIPPED_VERSION="::${lAPP_NAME//\.exe}:${lAPP_VERS:-NA}"
+
+  ### new SBOM json testgenerator
+  if command -v jo >/dev/null; then
```
Comment (BenediktMKuehne) [inline modules/S08_main_package_sbom_modules/S08_submodule_windows_exifparser.sh:191]: .

**Review by BenediktMKuehne — COMMENTED** [review]
Just some questions

File: modules/S08_main_package_sbom_modules/S08_submodule_deb_package_parser.sh (hunk 1 above)
Comment (m-1-k-3) [reply modules/S08_main_package_sbom_modules/S08_submodule_deb_package_parser.sh:138]: now it is ... I will remove this check soon

**Pull request merged**